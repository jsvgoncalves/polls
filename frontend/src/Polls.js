import React from 'react';
import { Row, Col } from 'react-grid-system';
import {API} from './API';
import Search from './Search';


//!TODO: Fix hardcoded links.
//       Possibly extract it to API and have only `/polls` here?
//       or make a `getAllPolls()` call
//!TODO: Do something more appropriate then just
//       logging the error.
class Polls extends React.Component {
  constructor (props) {
    super(props);
    this.state = {items: []}
    this.updateItems = this.updateItems.bind(this);
    this.searchUpdate = this.searchUpdate.bind(this);
  }
  updateItems (items) {
    this.setState({items})
  }
  componentWillMount () {
    API.fetch(`${this.props.url}/polls`)
       .then(({_items: items}) => this.updateItems(items))
       .catch((e) => console.log(e));
  }
  searchUpdate(s) {
    // "title": {"$regex": ".*a.*"}
    const query = `${this.props.url}/polls?where={"title": {"$regex": ".*${s}.*"}}`;
    API.fetch(query)
       .then(({_items: items}) => this.updateItems(items))
       .catch((e) => console.log(e));
  }
  render () {
    let items = this.state.items;
    return (
      <div>
      <Row>
        <Col xs={12} md={12}><Search update={this.searchUpdate.bind(this)}/></Col>
      </Row>
      {items.map(item =>
        (<Row key={item._id} className="poll">
          <Col><Poll poll={item} /></Col>
        </Row>)
      )}
      </div>
    );
  }
}

const Poll = (props) => (
  <div>
    <Row>
      <Col className="poll-field"><strong>{props.poll.title}</strong></Col>
    </Row>
    <Row>
      <Col className="poll-field" xs={12} md={6}>{props.poll.initiator.name}</Col>
      <Col className="poll-field" xs={12} offset={{ md: 1 }} md={5}>{props.poll._created}</Col>
    </Row>
    <Row>
      <Col className="poll-field">{props.poll.description || <em>No description provided</em>}</Col>
    </Row>
  </div>);

export default Polls;
