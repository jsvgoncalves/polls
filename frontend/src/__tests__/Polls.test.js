import React from 'react';
import ReactDOM from 'react-dom';
import {mount} from 'enzyme'
import toJson from 'enzyme-to-json';
import Polls from '../Polls'

const polls = {
  _items: [
    {
      title: 'Poll 1',
      _id: 'poll1',
      initiator: {name: 'John Doe'},
      _created: 'date'
    },
    {
      title: 'Poll 2',
      _id: 'poll2',
      initiator: {name: 'John Doe'},
      _created: 'date'
    }
    ]
};

describe('Polls', () => {
  it('renders correctly', () => {
    const pp = mount(
      <Polls />
    );
    pp.instance().updateItems(polls._items)
    expect(toJson(pp)).toMatchSnapshot();
  });

  it('requests updated polls', () => {
    global.fetch = jest.fn().mockImplementation(() => {
        return new Promise((resolve, reject) => {
          resolve({
            // The actual response
            _id: '123',
            // HTTP Status Code
            status: 200,
            // The response needs to have a .json() method
            json: () => { return polls }
          });
        });
    });
    expect.assertions(1);
    const div = document.createElement('div');
    ReactDOM.render(<Polls />, div);
    expect(fetch).toHaveBeenCalled();
  });
});
