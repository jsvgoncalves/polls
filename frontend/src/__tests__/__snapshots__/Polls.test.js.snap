exports[`Polls renders correctly 1`] = `
<Polls>
  <div>
    <Row
      style={Object {}}>
      <div
        style={
          Object {
            "marginLeft": "-15px",
            "marginRight": "-15px",
          }
        }>
        <Col
          lg={12}
          md={12}
          offset={Object {}}
          sm={12}
          style={Object {}}
          xl={12}
          xs={12}>
          <div
            style={
              Object {
                "boxSizing": "border-box",
                "float": "left",
                "marginLeft": "0%",
                "minHeight": "1px",
                "paddingLeft": "15px",
                "paddingRight": "15px",
                "width": "100%",
              }
            }>
            <Search
              update={[Function]}>
              <Row
                className="search-bar"
                style={Object {}}>
                <div
                  className="search-bar"
                  style={
                    Object {
                      "marginLeft": "-15px",
                      "marginRight": "-15px",
                    }
                  }>
                  <form
                    onSubmit={[Function]}>
                    <Col
                      className="search-col"
                      lg={12}
                      md={6}
                      offset={
                        Object {
                          "md": 4,
                        }
                      }
                      sm={12}
                      style={Object {}}
                      xl={12}
                      xs={12}>
                      <div
                        className="search-col"
                        style={
                          Object {
                            "boxSizing": "border-box",
                            "float": "left",
                            "marginLeft": "33.333333333333336%",
                            "minHeight": "1px",
                            "paddingLeft": "15px",
                            "paddingRight": "15px",
                            "width": "50%",
                          }
                        }>
                        <input
                          onChange={[Function]}
                          placeholder="Search by title"
                          value="" />
                      </div>
                    </Col>
                    <Col
                      className="search-col"
                      lg={12}
                      md={2}
                      offset={Object {}}
                      sm={12}
                      style={Object {}}
                      xl={12}
                      xs={12}>
                      <div
                        className="search-col"
                        style={
                          Object {
                            "boxSizing": "border-box",
                            "float": "left",
                            "marginLeft": "0%",
                            "minHeight": "1px",
                            "paddingLeft": "15px",
                            "paddingRight": "15px",
                            "width": "16.666666666666668%",
                          }
                        }>
                        <button
                          type="submit">
                          Search
                        </button>
                      </div>
                    </Col>
                  </form>
                  <ClearFix
                    lg={true}
                    md={true}
                    sm={true}
                    xl={true}
                    xs={true}>
                    <Visible
                      lg={true}
                      md={true}
                      sm={true}
                      xl={true}
                      xs={true}>
                      <RenderAny>
                        <div
                          style={
                            Object {
                              "clear": "both",
                              "display": "table",
                            }
                          }>
                          <span
                            style={
                              Object {
                                "display": "none",
                              }
                            }>
                             
                          </span>
                        </div>
                      </RenderAny>
                    </Visible>
                  </ClearFix>
                </div>
              </Row>
            </Search>
          </div>
        </Col>
        <ClearFix
          lg={true}
          md={true}
          sm={true}
          xl={true}
          xs={true}>
          <Visible
            lg={true}
            md={true}
            sm={true}
            xl={true}
            xs={true}>
            <RenderAny>
              <div
                style={
                  Object {
                    "clear": "both",
                    "display": "table",
                  }
                }>
                <span
                  style={
                    Object {
                      "display": "none",
                    }
                  }>
                   
                </span>
              </div>
            </RenderAny>
          </Visible>
        </ClearFix>
      </div>
    </Row>
    <Row
      className="poll"
      style={Object {}}>
      <div
        className="poll"
        style={
          Object {
            "marginLeft": "-15px",
            "marginRight": "-15px",
          }
        }>
        <Col
          lg={12}
          md={12}
          offset={Object {}}
          sm={12}
          style={Object {}}
          xl={12}
          xs={12}>
          <div
            style={
              Object {
                "boxSizing": "border-box",
                "float": "left",
                "marginLeft": "0%",
                "minHeight": "1px",
                "paddingLeft": "15px",
                "paddingRight": "15px",
                "width": "100%",
              }
            }>
            <Poll
              poll={
                Object {
                  "_created": "date",
                  "_id": "poll1",
                  "initiator": Object {
                    "name": "John Doe",
                  },
                  "title": "Poll 1",
                }
              }>
              <div>
                <Row
                  style={Object {}}>
                  <div
                    style={
                      Object {
                        "marginLeft": "-15px",
                        "marginRight": "-15px",
                      }
                    }>
                    <Col
                      className="poll-field"
                      lg={12}
                      md={12}
                      offset={Object {}}
                      sm={12}
                      style={Object {}}
                      xl={12}
                      xs={12}>
                      <div
                        className="poll-field"
                        style={
                          Object {
                            "boxSizing": "border-box",
                            "float": "left",
                            "marginLeft": "0%",
                            "minHeight": "1px",
                            "paddingLeft": "15px",
                            "paddingRight": "15px",
                            "width": "100%",
                          }
                        }>
                        <strong>
                          Poll 1
                        </strong>
                      </div>
                    </Col>
                    <ClearFix
                      lg={true}
                      md={true}
                      sm={true}
                      xl={true}
                      xs={true}>
                      <Visible
                        lg={true}
                        md={true}
                        sm={true}
                        xl={true}
                        xs={true}>
                        <RenderAny>
                          <div
                            style={
                              Object {
                                "clear": "both",
                                "display": "table",
                              }
                            }>
                            <span
                              style={
                                Object {
                                  "display": "none",
                                }
                              }>
                               
                            </span>
                          </div>
                        </RenderAny>
                      </Visible>
                    </ClearFix>
                  </div>
                </Row>
                <Row
                  style={Object {}}>
                  <div
                    style={
                      Object {
                        "marginLeft": "-15px",
                        "marginRight": "-15px",
                      }
                    }>
                    <Col
                      className="poll-field"
                      lg={12}
                      md={6}
                      offset={Object {}}
                      sm={12}
                      style={Object {}}
                      xl={12}
                      xs={12}>
                      <div
                        className="poll-field"
                        style={
                          Object {
                            "boxSizing": "border-box",
                            "float": "left",
                            "marginLeft": "0%",
                            "minHeight": "1px",
                            "paddingLeft": "15px",
                            "paddingRight": "15px",
                            "width": "50%",
                          }
                        }>
                        John Doe
                      </div>
                    </Col>
                    <Col
                      className="poll-field"
                      lg={12}
                      md={5}
                      offset={
                        Object {
                          "md": 1,
                        }
                      }
                      sm={12}
                      style={Object {}}
                      xl={12}
                      xs={12}>
                      <div
                        className="poll-field"
                        style={
                          Object {
                            "boxSizing": "border-box",
                            "float": "left",
                            "marginLeft": "8.333333333333334%",
                            "minHeight": "1px",
                            "paddingLeft": "15px",
                            "paddingRight": "15px",
                            "width": "41.66666666666667%",
                          }
                        }>
                        date
                      </div>
                    </Col>
                    <ClearFix
                      lg={true}
                      md={true}
                      sm={true}
                      xl={true}
                      xs={true}>
                      <Visible
                        lg={true}
                        md={true}
                        sm={true}
                        xl={true}
                        xs={true}>
                        <RenderAny>
                          <div
                            style={
                              Object {
                                "clear": "both",
                                "display": "table",
                              }
                            }>
                            <span
                              style={
                                Object {
                                  "display": "none",
                                }
                              }>
                               
                            </span>
                          </div>
                        </RenderAny>
                      </Visible>
                    </ClearFix>
                  </div>
                </Row>
                <Row
                  style={Object {}}>
                  <div
                    style={
                      Object {
                        "marginLeft": "-15px",
                        "marginRight": "-15px",
                      }
                    }>
                    <Col
                      className="poll-field"
                      lg={12}
                      md={12}
                      offset={Object {}}
                      sm={12}
                      style={Object {}}
                      xl={12}
                      xs={12}>
                      <div
                        className="poll-field"
                        style={
                          Object {
                            "boxSizing": "border-box",
                            "float": "left",
                            "marginLeft": "0%",
                            "minHeight": "1px",
                            "paddingLeft": "15px",
                            "paddingRight": "15px",
                            "width": "100%",
                          }
                        }>
                        <em>
                          No description provided
                        </em>
                      </div>
                    </Col>
                    <ClearFix
                      lg={true}
                      md={true}
                      sm={true}
                      xl={true}
                      xs={true}>
                      <Visible
                        lg={true}
                        md={true}
                        sm={true}
                        xl={true}
                        xs={true}>
                        <RenderAny>
                          <div
                            style={
                              Object {
                                "clear": "both",
                                "display": "table",
                              }
                            }>
                            <span
                              style={
                                Object {
                                  "display": "none",
                                }
                              }>
                               
                            </span>
                          </div>
                        </RenderAny>
                      </Visible>
                    </ClearFix>
                  </div>
                </Row>
              </div>
            </Poll>
          </div>
        </Col>
        <ClearFix
          lg={true}
          md={true}
          sm={true}
          xl={true}
          xs={true}>
          <Visible
            lg={true}
            md={true}
            sm={true}
            xl={true}
            xs={true}>
            <RenderAny>
              <div
                style={
                  Object {
                    "clear": "both",
                    "display": "table",
                  }
                }>
                <span
                  style={
                    Object {
                      "display": "none",
                    }
                  }>
                   
                </span>
              </div>
            </RenderAny>
          </Visible>
        </ClearFix>
      </div>
    </Row>
    <Row
      className="poll"
      style={Object {}}>
      <div
        className="poll"
        style={
          Object {
            "marginLeft": "-15px",
            "marginRight": "-15px",
          }
        }>
        <Col
          lg={12}
          md={12}
          offset={Object {}}
          sm={12}
          style={Object {}}
          xl={12}
          xs={12}>
          <div
            style={
              Object {
                "boxSizing": "border-box",
                "float": "left",
                "marginLeft": "0%",
                "minHeight": "1px",
                "paddingLeft": "15px",
                "paddingRight": "15px",
                "width": "100%",
              }
            }>
            <Poll
              poll={
                Object {
                  "_created": "date",
                  "_id": "poll2",
                  "initiator": Object {
                    "name": "John Doe",
                  },
                  "title": "Poll 2",
                }
              }>
              <div>
                <Row
                  style={Object {}}>
                  <div
                    style={
                      Object {
                        "marginLeft": "-15px",
                        "marginRight": "-15px",
                      }
                    }>
                    <Col
                      className="poll-field"
                      lg={12}
                      md={12}
                      offset={Object {}}
                      sm={12}
                      style={Object {}}
                      xl={12}
                      xs={12}>
                      <div
                        className="poll-field"
                        style={
                          Object {
                            "boxSizing": "border-box",
                            "float": "left",
                            "marginLeft": "0%",
                            "minHeight": "1px",
                            "paddingLeft": "15px",
                            "paddingRight": "15px",
                            "width": "100%",
                          }
                        }>
                        <strong>
                          Poll 2
                        </strong>
                      </div>
                    </Col>
                    <ClearFix
                      lg={true}
                      md={true}
                      sm={true}
                      xl={true}
                      xs={true}>
                      <Visible
                        lg={true}
                        md={true}
                        sm={true}
                        xl={true}
                        xs={true}>
                        <RenderAny>
                          <div
                            style={
                              Object {
                                "clear": "both",
                                "display": "table",
                              }
                            }>
                            <span
                              style={
                                Object {
                                  "display": "none",
                                }
                              }>
                               
                            </span>
                          </div>
                        </RenderAny>
                      </Visible>
                    </ClearFix>
                  </div>
                </Row>
                <Row
                  style={Object {}}>
                  <div
                    style={
                      Object {
                        "marginLeft": "-15px",
                        "marginRight": "-15px",
                      }
                    }>
                    <Col
                      className="poll-field"
                      lg={12}
                      md={6}
                      offset={Object {}}
                      sm={12}
                      style={Object {}}
                      xl={12}
                      xs={12}>
                      <div
                        className="poll-field"
                        style={
                          Object {
                            "boxSizing": "border-box",
                            "float": "left",
                            "marginLeft": "0%",
                            "minHeight": "1px",
                            "paddingLeft": "15px",
                            "paddingRight": "15px",
                            "width": "50%",
                          }
                        }>
                        John Doe
                      </div>
                    </Col>
                    <Col
                      className="poll-field"
                      lg={12}
                      md={5}
                      offset={
                        Object {
                          "md": 1,
                        }
                      }
                      sm={12}
                      style={Object {}}
                      xl={12}
                      xs={12}>
                      <div
                        className="poll-field"
                        style={
                          Object {
                            "boxSizing": "border-box",
                            "float": "left",
                            "marginLeft": "8.333333333333334%",
                            "minHeight": "1px",
                            "paddingLeft": "15px",
                            "paddingRight": "15px",
                            "width": "41.66666666666667%",
                          }
                        }>
                        date
                      </div>
                    </Col>
                    <ClearFix
                      lg={true}
                      md={true}
                      sm={true}
                      xl={true}
                      xs={true}>
                      <Visible
                        lg={true}
                        md={true}
                        sm={true}
                        xl={true}
                        xs={true}>
                        <RenderAny>
                          <div
                            style={
                              Object {
                                "clear": "both",
                                "display": "table",
                              }
                            }>
                            <span
                              style={
                                Object {
                                  "display": "none",
                                }
                              }>
                               
                            </span>
                          </div>
                        </RenderAny>
                      </Visible>
                    </ClearFix>
                  </div>
                </Row>
                <Row
                  style={Object {}}>
                  <div
                    style={
                      Object {
                        "marginLeft": "-15px",
                        "marginRight": "-15px",
                      }
                    }>
                    <Col
                      className="poll-field"
                      lg={12}
                      md={12}
                      offset={Object {}}
                      sm={12}
                      style={Object {}}
                      xl={12}
                      xs={12}>
                      <div
                        className="poll-field"
                        style={
                          Object {
                            "boxSizing": "border-box",
                            "float": "left",
                            "marginLeft": "0%",
                            "minHeight": "1px",
                            "paddingLeft": "15px",
                            "paddingRight": "15px",
                            "width": "100%",
                          }
                        }>
                        <em>
                          No description provided
                        </em>
                      </div>
                    </Col>
                    <ClearFix
                      lg={true}
                      md={true}
                      sm={true}
                      xl={true}
                      xs={true}>
                      <Visible
                        lg={true}
                        md={true}
                        sm={true}
                        xl={true}
                        xs={true}>
                        <RenderAny>
                          <div
                            style={
                              Object {
                                "clear": "both",
                                "display": "table",
                              }
                            }>
                            <span
                              style={
                                Object {
                                  "display": "none",
                                }
                              }>
                               
                            </span>
                          </div>
                        </RenderAny>
                      </Visible>
                    </ClearFix>
                  </div>
                </Row>
              </div>
            </Poll>
          </div>
        </Col>
        <ClearFix
          lg={true}
          md={true}
          sm={true}
          xl={true}
          xs={true}>
          <Visible
            lg={true}
            md={true}
            sm={true}
            xl={true}
            xs={true}>
            <RenderAny>
              <div
                style={
                  Object {
                    "clear": "both",
                    "display": "table",
                  }
                }>
                <span
                  style={
                    Object {
                      "display": "none",
                    }
                  }>
                   
                </span>
              </div>
            </RenderAny>
          </Visible>
        </ClearFix>
      </div>
    </Row>
  </div>
</Polls>
`;
